builder: Alias `image` to `moby` exporter

vvoland opened the pull request:
Moby exporter is a wrapper for the `image` exporter that modifies the
request attributes on the fly to make it work correctly with the engine.

Before this, if user explicitly asks for the `image` exporter, it would
completely bypass the special handling it needs. For example it wouldn't
create an image if the user didn't provide a tag.

This makes the `image` exporter being handled the same way as `moby.

**- What I did**
- Alias `image` to `moby` exporter

**- How I did it**

**- How to verify it**

**- Description for the changelog**


**- A picture of a cute animal (not mandatory but encouraged)**

vvoland committed:
builder/worker: Handle image exporter same as moby

Moby exporter is a wrapper for the `image` exporter that modifies the
request attributes on the fly to make it work correctly with the engine.

Before this, if user explicitly asks for the `image` exporter, it would
completely bypass the special handling it needs. For example it wouldn't
create an image if the user didn't provide a tag.

This makes the `image` exporter being handled the same way as `moby.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

rumpl commented on builder/builder-next/worker/containerdworker.go:
```diff
@@ -28,7 +28,7 @@ func NewContainerdWorker(ctx context.Context, wo base.WorkerOpt) (*ContainerdWor
 func (w *ContainerdWorker) Exporter(name string, sm *session.Manager) (exporter.Exporter, error) {
 	switch name {
-	case mobyexporter.Moby:
+	case mobyexporter.Moby, client.ExporterImage:
```
This will force unpack (and other things we force with the wrapper) to any build request with the image exporter, I’m not sure we want to do that

vvoland replied:
What would be the the use case for using plain `image` with moby?

vvoland replied:
That's kind of my perspective too - `moby` exporter is just an implementation detail that allows us to override some options that make more sense in our context. It shouldn't really be visible on the outside and from the outside it would still be the `image` exporter.

rumpl replied:
My worry is that, with this change, the `image` exporter doesn't work as documented, since we force things.

And if moby does implement the `image` exporter then we can end up being able to build an image (inside our own `moby` containerd namespace) that's not visible by docker (because it's not unpacked for example), or can very well break docker if we don't handle each and every corner case.

neersighted replied:
We've established that they should (in large part) be equivalent after talking with @tonistiigi; the issue is more where the should should live in the long-term (https://github.com/moby/moby/pull/45666#pullrequestreview-1456184882).

For now, it's probably best to create an issue on BuildKit to move the wrapper functionality to that repo, and put a `TODO:` here linking to it.

tonistiigi replied:
> In my mind, moby is the exporter that will get you an image inside moby

`type=docker` exporter is what gets image inside `docker`, even if building inside container/k8s etc. `type=moby` is not even documented anywhere https://docs.docker.com/engine/reference/commandline/buildx_build/#output . It's an internal switch that `buildx` does because non-containerd moby is incapable of using the image exporter directly.

vvoland commented:
Related failure:
```
     client_test.go:1116: 
        	Error Trace:	/src/client/client_test.go:1116
        	Error:      	Received unexpected error:
        	            	invalid reference format
        	            	github.com/moby/buildkit/util/stack.Enable
        	            		/src/util/stack/stack.go:77
        	            	github.com/moby/buildkit/util/grpcerrors.FromGRPC
        	            		/src/util/grpcerrors/grpcerrors.go:197
        	            	github.com/moby/buildkit/util/grpcerrors.UnaryClientInterceptor
        	            		/src/util/grpcerrors/intercept.go:41
        	            	google.golang.org/grpc.(*ClientConn).Invoke
        	            		/src/vendor/google.golang.org/grpc/call.go:35
        	            	github.com/moby/buildkit/api/services/control.(*controlClient).Solve
        	            		/src/api/services/control/control.pb.go:2208
        	            	github.com/moby/buildkit/client.(*Client).solve.func2
        	            		/src/client/solve.go:258
        	            	golang.org/x/sync/errgroup.(*Group).Go.func1
        	            		/src/vendor/golang.org/x/sync/errgroup/errgroup.go:75
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1594
        	            	failed to solve
        	            	github.com/moby/buildkit/client.(*Client).solve.func2
        	            		/src/client/solve.go:273
        	            	golang.org/x/sync/errgroup.(*Group).Go.func1
        	            		/src/vendor/golang.org/x/sync/errgroup/errgroup.go:75
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1594
        	Test:       	TestIntegration/TestFrontendImageNaming/worker=dockerd-containerd/frontend/image

```

https://github.com/moby/buildkit/blob/v0.11.6/client/client_test.go#L1106

It fails because `image` is aliased to `moby` now and we don't handle the `*` special value in our wrapper and try to parse it as a normal reference.
We could handle it easily, but... perhaps we shouldn't: https://github.com/moby/moby/pull/45666#discussion_r1212393311

neersighted reviewed (changes requested):
@tonistiigi and I discussed this at painful length in the maintainer's call. Takeaways were:
* This PR is correct as the most pragmatic solution.
* The first change (to use the `moby` exporter for the REST API) is always correct: the 'legacy' behavior should apply here.
* The second change (to use the exporter wrapper even with the `image` exporter) is likely somewhat correct.
  * Today, an older Buildx with `type=image` or `type=docker` will get the same behavior pre-containerd.
  * However, with contianerd, the image will be 'lost' if not tagged when `type=image`.
* However, this behavior doesn't really belong in Moby; the logic to 'tag the image if no tag is provided so it's not discarded' should likely be part of BuildKit itself.
  * The test failure is a prime illustration of why this is the case; the wrapper is naive and very difficult to maintain if it's not part of BuildKit proper.
* We need to take this PR, with the addition of fixing the test failure/making sure the wrapper is resilient to more inputs.

vvoland committed:
builder/overrides: Handle "*" image name

Don't try to parse it as a comma separated list of references and
pass it through to the image exporter instead.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

rumpl commented on builder/builder-next/exporter/overrides/wrapper.go:
```diff
@@ -30,11 +30,17 @@ func (e *imageExporterMobyWrapper) Resolve(ctx context.Context, exporterAttrs ma
-		return nil, err
+	if name, ok := exporterAttrs[keyImageName]; ok {
+		if name == "*" {
+			exporterAttrs[keyImageName] = "*"
```
This `exporterAttrs[keyImageName]` already equals to `*` so this is not needed

vvoland commented:
Failure somewhat related:
```
     client_test.go:817: 
        	Error Trace:	/src/client/client_test.go:817
        	            				/src/client/run.go:87
        	            				/src/client/run.go:196
        	Error:      	Received unexpected error:
        	            	failed to push localhost:42703/foo/bar:latest: can't push tagged ref localhost:42703/foo/bar:latest by digest
        	            	github.com/moby/buildkit/util/stack.Enable
        	            		/src/util/stack/stack.go:77
        	            	github.com/moby/buildkit/util/grpcerrors.FromGRPC
        	            		/src/util/grpcerrors/grpcerrors.go:197
        	            	github.com/moby/buildkit/util/grpcerrors.UnaryClientInterceptor
        	            		/src/util/grpcerrors/intercept.go:41
        	            	google.golang.org/grpc.(*ClientConn).Invoke
        	            		/src/vendor/google.golang.org/grpc/call.go:35
        	            	github.com/moby/buildkit/api/services/control.(*controlClient).Solve
        	            		/src/api/services/control/control.pb.go:2208
        	            	github.com/moby/buildkit/client.(*Client).solve.func2
        	            		/src/client/solve.go:258
        	            	golang.org/x/sync/errgroup.(*Group).Go.func1
        	            		/src/vendor/golang.org/x/sync/errgroup/errgroup.go:75
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1594
        	            	failed to solve
        	            	github.com/moby/buildkit/client.(*Client).solve.func2
        	            		/src/client/solve.go:273
        	            	golang.org/x/sync/errgroup.(*Group).Go.func1
        	            		/src/vendor/golang.org/x/sync/errgroup/errgroup.go:75
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1594
        	Test:       	TestIntegration/TestPushByDigest/worker=dockerd-containerd
```

our wrapper forces the names to have the `:latest` tag if it wasn't specified, but then it fails a check in buildkit: https://github.com/moby/buildkit/blob/80b91c55a47feb52a66e143a3929e614dd5c7adc/util/push/push.go#L58-L60

neersighted commented:
@tonistiigi @crazy-max Should we still prefer moving the logic to BuildKit over this PR?
